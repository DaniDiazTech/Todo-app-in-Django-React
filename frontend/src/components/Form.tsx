import React, { useContext, useState } from 'react';
import { sendData } from '../helpers/sendData';
import { fieldsInterface, taskInterface } from '../interfaces/appInterfaces';
import { getData } from '../helpers/getData';
import { ToDoContext } from '../App';

const Form = () => {
  const { setTaskList } = useContext(ToDoContext);
  const [task, setTask] = useState<taskInterface>({
    id: null,
    nameTask: '',
    description: '',
    completed: false,
    editing: false,
  });

  const handleInputChange = (e: any) => {
    const { name, value }: fieldsInterface = e.target;
    setTask({
      ...task,
      [name]: value,
    });
  };

  const handleSubmit = async (e: any) => {
    e.preventDefault();
    await sendData('http://localhost:8000/api/task-create/', task);
    setTask({
      id: null,
      nameTask: '',
      description: '',
      completed: false,
      editing: false,
    });
    getData().then(data => setTaskList({
      data,
      loading: false
    }))
  };

  return (
    <form className="container-fluid" onSubmit={handleSubmit}>
      <div className="mb-3">
        <input
          type="text"
          name="nameTask"
          className="form-control"
          placeholder="Task name"
          value={task.nameTask}
          onChange={handleInputChange}
        />
      </div>
      <div className="mb-3">
        <textarea
          className="form-control"
          name="description"
          placeholder="Task description"
          value={task.description}
          onChange={handleInputChange}
        ></textarea>
      </div>
      <button className="btn btn-primary">Submit</button>
    </form>
  );
};

export default React.memo(Form);
