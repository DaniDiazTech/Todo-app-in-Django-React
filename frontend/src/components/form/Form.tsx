import React, { useContext } from 'react';
import { sendData } from '../../helpers/sendData/sendData';
import { fieldsInterface } from '../../interfaces/appInterfaces';
import { getData } from '../../helpers/getData/getData';
import { ToDoContext } from '../../context/ToDoProvider';

const Form = () => {
  const { setTaskList,task,setTask } = useContext(ToDoContext);
  
  const handleInputChange = (e: any) => {
    const { name, value }: fieldsInterface = e.target;
    setTask({
      ...task,
      [name]: value,
    });
  };

  const handleSubmit = async (e: any) => {
    e.preventDefault();
    
    let url = ""
    let action = "";

    if(task.editing){
      url = `http://localhost:8000/api/task-update/${task.id}/`
      action = 'PUT'
      await sendData(url, task, action);
    }else {
       url = 'http://localhost:8000/api/task-create/'
       action = 'POST';
       await sendData(url, task, action);
    }
  
    setTask({
      id: null,
      nameTask: '',
      description: '',
      completed: false,
      editing: false,
    });

    getData().then(data => setTaskList({
      data,
      loading: false
    }))
  };

  return (
    <form className="container-fluid" onSubmit={ handleSubmit }>
      <div className="mb-3">
        <input
          type="text"
          name="nameTask"
          className="form-control"
          placeholder="Task name"
          value={task.nameTask}
          onChange={handleInputChange}
        />
      </div>
      <div className="mb-3">
        <textarea
          className="form-control"
          name="description"
          placeholder="Task description"
          value={task.description}
          onChange={handleInputChange}
        ></textarea>
      </div>
      <button className="btn btn-primary">{task.editing ? 'Edit'  : 'Submit'}</button>
    </form>
  );
};

export default React.memo(Form);
