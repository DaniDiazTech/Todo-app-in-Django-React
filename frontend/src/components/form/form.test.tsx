import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import Form from './Form';

describe('Test on <Form />', () => {
  const wrapper = shallow(<Form />);

  test('Should render component', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('Should input change', () => {
    const value = 'Value';
    const input = wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value } });

    expect(input.prop('value')).toBe(value);
  });

  test('Should send data', () => {});
});
