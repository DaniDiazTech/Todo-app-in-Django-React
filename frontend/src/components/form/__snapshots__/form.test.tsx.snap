// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <Form /> Should render component 1`] = `
<form
  className="container-fluid"
  onSubmit={[Function]}
>
  <div
    className="form-outline mb-4"
  >
    <input
      className="form-control"
      name="nameTask"
      onChange={[Function]}
      placeholder="Task name"
      value=""
    />
  </div>
  <div
    className="form-outline mb-4"
  >
    <input
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Task description"
      type="text"
      value=""
    />
  </div>
  <button
    className="btn btn-primary btn-block"
    type="submit"
  >
    Submit
  </button>
</form>
`;
